### PR TITLE
feat: Add support for the API version header

### PR DESCRIPTION
**Pull request by Hectorhammett**
Added support for the API version header from the new Versioning header for apis.

### CODE REVIEW

File: src/Generation/GapicClientV2Generator.php
Comment (Hectorhammett) [inline src/Generation/GapicClientV2Generator.php:173]: I gave this a thought and I don't love how I am just sending a string.

I was wondering that using something similar like the line above with `Access::PRIVATE` would look better, but I wonder how we can handle different classes, specially ones that we create. Of course this is with the thought that we should support any class in order to keep it generic.

Perhaps we could create an Enum with just the basic classes for our needs?

Comment (noahdietz) [reply src/Generation/GapicClientV2Generator.php:173]: Like `AST::TYPE_STRING`, `AST::TYPE_ARRAY`, etc. ? Sure that makes sense to me. I agree "magic values" aren't as fun as "magic" might make them seem.

I wouldn't worry about adding constants for each known possible type just yet. Better to just add them as we need them.

Comment (Hectorhammett) [reply src/Generation/GapicClientV2Generator.php:173]: Yeah something like that.

I am gonna add that then. I like the idea of using the types we need as we need them 👍

Comment (Hectorhammett) [reply src/Generation/GapicClientV2Generator.php]: I ended up creating a class inside the `hasType` file to keep everything tidy as in the `Access` file. The class that contains all the constants is called `TypeDeclaration`.

**Review by noahdietz — APPROVED** [review]
This is tight tight hella tight. Well done Hector

File: src/Ast/HasType.php
Comment (bshaffer) [inline src/Ast/HasType.php:26]: This should be `HasTypeTrait`

File: src/Ast/HasType.php
Comment (bshaffer) [inline src/Ast/HasType.php:21]: This should be in a separate class file, and we can add other constants to it there.

**Review by bshaffer — CHANGES_REQUESTED** [review]

File: src/Ast/HasType.php
Comment (Hectorhammett) [reply src/Ast/HasType.php:26]: I am currently following the convention that is on the `PhpClassMember` file. It has 3 traits:
image

I would like to keep things nice and tidy. Do we vote for updating all the traits to call them `trait`?

File: src/Ast/HasType.php
Comment (Hectorhammett) [reply src/Ast/HasType.php:21]: Good catch! 

I updated my code to make use of `ResolvedType`. Should we remove this one then? I vote yes.

Comment (Hectorhammett) [reply src/Ast/HasType.php]: I decided to remove this one in favor for the `ResolvedType` type

File: src/Ast/HasType.php
Comment (bshaffer) [reply src/Ast/HasType.php:23]: Since the precedent has already been set, we can keep them named as-is. Thanks for pointing that out.

File: src/Ast/PhpClassMember.php
Comment (bshaffer) [inline src/Ast/PhpClassMember.php:26]: I would suggest moving this to just `PhpProperty`, as the other three classes which extend this class (`PhpConstant`, `PhpComment`, and `PhpMethod`) don't support types in the same way:

 - Constants: defining types are invalid (since they're implicit from the constant value)
 - Comments: don't have types
 - Methods: types are already defined by parameters return types

And we probably want to just get rid of the trait all together, since traits are meant for reusing code, and we will no longer be doing that.

Comment (Hectorhammett) [reply src/Ast/PhpClassMember.php:26]: Great comment. Makes sense, let me change it.

File: src/Ast/PhpProperty.php
Comment (bshaffer) [inline src/Ast/PhpProperty.php:30]: As of PHP 8.2, [setting dynamic properties is deprecated](https://php.watch/versions/8.2/dynamic-properties-deprecated). This repo is currently on PHP 8.1, which is why no errors are thrown. We can wait until we update the repo to PHP 8.2 to add these properties, or we could add it now, e.g. 

```php
final class PhpProperty extends PhpClassMember
{
    private ?ResolvedType $type;
```

... but since we'll have to update these across the whole library eventually, I'll leave that up to you whether you'd like to define it now or not.

File: tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php
Comment (bshaffer) [inline tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php:89]: woops 😦

**Review by bshaffer — APPROVED** [review]
Looks great! Just one suggestion

File: src/Ast/PhpProperty.php
Comment (Hectorhammett) [reply src/Ast/PhpProperty.php:30]: Ehhh let's change it once and for all.

File: tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php
Comment (Hectorhammett) [reply tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php:89]: I rebased with main and this should be the correct name isn't it? The proto defines it with lower case. No idea why it changed now.

File: src/Ast/PhpProperty.php
Comment (bshaffer) [reply src/Ast/PhpProperty.php:30]: If it's between introducing a "consistent" change that goes against a best practice, and introducing an "inconsistent" change that follows best practices, I'd prefer to go with the latter

File: tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php
Comment (bshaffer) [reply tests/Unit/ProtoTests/BasicGrpcOnly/out/src/Client/BasicGrpcOnlyClient.php:89]: I meant woops because I introduced this in the previous PR by changing it to lowercase before merging. This would have been fine, but I forgot to update the fixtures (as you've done here).

I'd say it's mostly the fault that the tests can pass even when they're accurate, because they aggressively cache.

File: src/Ast/PhpProperty.php
Comment (bshaffer) [reply src/Ast/PhpProperty.php:30]: ...and I think having undefined properties is a really bad practice

Comment (Hectorhammett) [reply src/Ast/PhpProperty.php:30]: Woops, I have this change but did not included the file on the commit. Will fix it rn.

Comment (Hectorhammett) [reply src/Ast/PhpProperty.php:30]: Woops, I have this change but didn't include the file on the Commit. Fixing it now.

File: src/Ast/PhpProperty.php
Comment (Hectorhammett) [inline src/Ast/PhpProperty.php:26]: Added the property here, I resolved the conversation below.

**Pull request merged**
